File: frontend/kube-guard-frontend/src/api/nodes.ts
// src/api/nodes.ts
import { get, post } from './http'

export type NodeRow = {
  name: string
  ip: string
  role: 'control-plane' | 'worker'
  status: 'Ready' | 'NotReady'
  unschedulable?: boolean
  kubeletVersion: string
  osImage: string
  kernelVersion: string
  containerRuntime: string
  cpuTotal: number
  cpuUsed: number
  memTotal: number
  memUsed: number
  podCapacity: number
  podUsed: number
}

export function fetchNodes() {
  return get<NodeRow[]>('/api/nodes')
}

export type JoinInfoResp =
  | {
      type: 'k3s'
      ok: true
      command: string
      params?: {
        node_ip?: string | null
        flannel_iface?: string | null
        k3s_version?: string | null
      }
    }
  | {
      type: 'k3s'
      ok: false
      hint: string
      command_template: string
    }

export interface FetchJoinInfoParams {
  server?: string
  token?: string
  node_ip?: string
  flannel_iface?: string
  k3s_version?: string
}

export function fetchJoinInfo(params: FetchJoinInfoParams) {
  const sp = new URLSearchParams()

  if (params.server) sp.set('server', params.server)
  if (params.token) sp.set('token', params.token)
  if (params.node_ip) sp.set('node_ip', params.node_ip)
  if (params.flannel_iface) sp.set('flannel_iface', params.flannel_iface)
  if (params.k3s_version) sp.set('k3s_version', params.k3s_version)

  const qs = sp.toString()
  return get<JoinInfoResp>(`/api/nodes/join-info${qs ? `?${qs}` : ''}`)
}


export type RemoveNodeResp = {
  ok: boolean
  steps?: any[]
  hint?: string
  node_side_hint?: string[]
  detail?: any
}


export function removeNode(
  nodeName: string,
  params?: { grace_seconds?: number; timeout_seconds?: number; force?: boolean }
) {
  const sp = new URLSearchParams()
  if (params?.grace_seconds != null) sp.set('grace_seconds', String(params.grace_seconds))
  if (params?.timeout_seconds != null) sp.set('timeout_seconds', String(params.timeout_seconds))
  if (params?.force != null) sp.set('force', params.force ? 'true' : 'false')
  const qs = sp.toString()
  return post<RemoveNodeResp>(`/api/nodes/${encodeURIComponent(nodeName)}/remove${qs ? `?${qs}` : ''}`, {})
}

export type OfflineNodeResp = {
  ok: boolean
  steps?: any[]
  hint?: string
}

export function offlineNode(
  nodeName: string,
  params?: { drain?: boolean; grace_seconds?: number; timeout_seconds?: number; force?: boolean }
) {
  const sp = new URLSearchParams()
  if (params?.drain != null) sp.set('drain', params.drain ? 'true' : 'false')
  if (params?.grace_seconds != null) sp.set('grace_seconds', String(params.grace_seconds))
  if (params?.timeout_seconds != null) sp.set('timeout_seconds', String(params.timeout_seconds))
  if (params?.force != null) sp.set('force', params.force ? 'true' : 'false')
  const qs = sp.toString()
  return post<OfflineNodeResp>(`/api/nodes/${encodeURIComponent(nodeName)}/offline${qs ? `?${qs}` : ''}`, {})
}

export type UncordonResp = { ok: boolean; node?: string; action?: string; detail?: any }

export function uncordonNode(nodeName: string) {
  return post<UncordonResp>(`/api/nodes/${encodeURIComponent(nodeName)}/uncordon`, {})
}

File: backend/kube-devops-backend/fastApiProject/routers/nodes.py
# routers/nodes.py
from __future__ import annotations

from fastapi import APIRouter, Query

from services.nodes.service import list_nodes
from services.nodes.ops import remove_node, offline_node, uncordon_node

router = APIRouter(prefix="/api/nodes", tags=["nodes"])


@router.post("/test")
def test_nodes_router():
    return {"ok": True, "msg": "nodes router is alive"}


@router.get("")
def get_nodes():
    """节点列表：GET /api/nodes"""
    return list_nodes()


@router.get("/join-info")
def get_join_info(
    server: str = Query("", description="K3S server url, e.g. https://1.2.3.4:6443"),
    token: str = Query("", description="K3S token"),

    # ✅ 可选：安装参数
    node_ip: str = Query("", description="--node-ip for k3s agent"),
    flannel_iface: str = Query("", description="--flannel-iface, e.g. eth0"),

    # ✅ 可选：固定版本（强烈建议填成和 master 一样）
    k3s_version: str = Query("", description='INSTALL_K3S_VERSION, e.g. "v1.33.3+k3s1"'),
):
    """
    生成 k3s join 命令（仅返回命令，不执行）
    - server/token 必填，否则返回模板
    - node_ip/flannel_iface 可选：生成 INSTALL_K3S_EXEC
    - k3s_version 可选：生成 INSTALL_K3S_VERSION，避免版本漂移
    """
    server = (server or "").strip()
    token = (token or "").strip()
    node_ip = (node_ip or "").strip()
    flannel_iface = (flannel_iface or "").strip()
    k3s_version = (k3s_version or "").strip()

    if not server or not token:
        return {
            "type": "k3s",
            "ok": False,
            "hint": "请提供 server 和 token（可选再填 node_ip / flannel_iface / k3s_version）",
            "command_template": (
                "curl -sfL https://rancher-mirror.rancher.cn/k3s/k3s-install.sh | "
                "INSTALL_K3S_MIRROR=cn "
                'INSTALL_K3S_VERSION="v1.33.3+k3s1" '
                "K3S_URL=<server> K3S_TOKEN=<token> "
                'INSTALL_K3S_EXEC="--node-ip=<node-ip> --flannel-iface=<iface>" '
                "sh -"
            ),
        }

    # ---- 组装 INSTALL_K3S_EXEC（可为空）----
    exec_args: list[str] = []
    if node_ip:
        exec_args.append(f"--node-ip={node_ip}")
    if flannel_iface:
        exec_args.append(f"--flannel-iface={flannel_iface}")

    install_exec_part = ""
    if exec_args:
        install_exec_part = f'INSTALL_K3S_EXEC="{" ".join(exec_args)}" '

    # ---- 组装 INSTALL_K3S_VERSION（可为空）----
    install_version_part = ""
    if k3s_version:
        install_version_part = f'INSTALL_K3S_VERSION="{k3s_version}" '

    cmd = (
        "curl -sfL https://rancher-mirror.rancher.cn/k3s/k3s-install.sh | "
        "INSTALL_K3S_MIRROR=cn "
        f"{install_version_part}"
        f"K3S_URL={server} "
        f"K3S_TOKEN={token} "
        f"{install_exec_part}"
        "sh -"
    )

    return {
        "type": "k3s",
        "ok": True,
        "command": cmd,
        "params": {
            "node_ip": node_ip or None,
            "flannel_iface": flannel_iface or None,
            "k3s_version": k3s_version or None,
        },
    }


@router.post("/{node_name}/offline")
def api_offline_node(
    node_name: str,
    drain: bool = Query(True),
    grace_seconds: int = Query(30, ge=0, le=600),
    timeout_seconds: int = Query(180, ge=30, le=1800),
    force: bool = Query(False),
):
    return offline_node(
        node_name=node_name,
        drain=drain,
        grace_seconds=grace_seconds,
        timeout_seconds=timeout_seconds,
        force=force,
    )


@router.post("/{node_name}/uncordon")
def api_uncordon_node(node_name: str):
    return uncordon_node(node_name)


@router.post("/{node_name}/remove")
def api_remove_node(
    node_name: str,
    grace_seconds: int = Query(30, ge=0, le=600),
    timeout_seconds: int = Query(180, ge=30, le=1800),
    force: bool = Query(False),
):
    return remove_node(
        node_name=node_name,
        grace_seconds=grace_seconds,
        timeout_seconds=timeout_seconds,
        force=force,
    )